src/data_chain/chain.rs

```diff
@@ -185,3 +171,309 @@ impl Chain {
+        payload = NodeState {
+            state: State::ElderLive,
+            public_key: *node_c.public_info().sign_key(),
+            age: node_c.public_info().age(),
+            section: prefix,
+        };
+        let live_c = tests::create_block(payload, &nodes, false);
+
+        // Create chain containing just Live(A).
+        let mut chain = Chain::from_blocks(vec![live_a.clone()], vec![], nodes.len());
+        let mut expected_blocks = vec![live_a.clone()];
+        let mut expected_pendings = vec![];
+
+        // Add Live(C) - should be held in `pending_blocks` as there's no corresponding Offline/
+        // Demoted/Relocated in the `pending_blocks`.
+        chain.add_block(live_c.clone());
+        expected_pendings.push(live_c.clone());
+        assert_eq!(chain.blocks, expected_blocks);
+        assert_eq!(chain.pending_blocks, expected_pendings);
+
+        // Add Relocated(B) - should be held in `pending_blocks` as B isn't marked as Live in the
+        // chain anywhere yet.
+        chain.add_block(relocated_b.clone());
+        expected_pendings.push(relocated_b.clone());
+        assert_eq!(chain.blocks, expected_blocks);
+        assert_eq!(chain.pending_blocks, expected_pendings);
+
+        // Add Live(B) - should be held in `pending_blocks` as the only possible Offline/Demoted/
+        // Relocated in the `pending_blocks` is Relocated(B), and that can't be added since B isn't
+        // marked as Live in the chain anywhere yet.
+        chain.add_block(live_b.clone());
+        expected_pendings.push(live_b.clone());
+        assert_eq!(chain.blocks, expected_blocks);
+        assert_eq!(chain.pending_blocks, expected_pendings);
+
+        // Add Offline(A) - should cause Offline(A)/Live(B) to be added as a pair (favouring Live(B)
+        // over Live(C) since there's a valid Offline/Demoted/Relocated for B, but not C), then also
+        // should cause Relocated(B)/Live(C) to get moved to the `blocks`, leaving `pending_blocks`
+        // empty.
+        chain.add_block(offline_a.clone());
+        expected_blocks = vec![live_a, offline_a, live_b, relocated_b, live_c];
+        assert_eq!(chain.blocks, expected_blocks);
+        assert!(chain.pending_blocks.is_empty());
+    }
+
+    #[ignore]
+    #[test]
+    fn add_vote() {
+        // Start with the chain having Live(A) in `blocks` and Live(B) in `pending_blocks`.  Add a
+        // vote for each and check the vote is added correctly.
+        let mut rng = SeededRng::thread_rng();
+        unwrap!(rust_sodium::init_with_rng(&mut rng));
+
+        let prefix = Prefix::default().pushed(rng.gen()).with_version(rng.gen());
+        let nodes = tests::create_full_infos(Some(prefix));
+
+        // Create Live(A) block and extra vote for Live(A).
+        let node_a = FullInfo::within_range(5, &prefix.lower_bound(), &prefix.upper_bound());
+        let mut payload = NodeState {
```

Comment:
Maybe we can reduce the verbosity of these tests by adding a `NodeState` constructor:
```rust
pub fn new(state: State, public_info: &PublicInfo, prefix: Prefix) -> Self
```
.